README.md
![shell screenshot with logo](/doc/icon-left-font-monochrome-black.svg)
> Wireguard Ligase is a collection of bash scripts designed to easily deploy Wireguard server and configure multiple clients in one go.

The main goals of Wireguard Ligase are:
- To make it easy for everyone to deploy and configure a Wireguard server and clients.
- To provide a platform for me to learn bash.
- To show that deploying your own VPN can be as easy as that.

For a comprehensive overview and guide, check out the [**Wiki**](https://github.com/SirToffski/WireGuard-Ligase/wiki).

Table of Contents
=================

  * [Using the script](#using-the-script)
     * [Pre-requisites](#pre-requisites)
     * [Usage](#usage)
  * [Why would I use this script?](#why-use-wireguard-ligase)
  * [Changelog](#changelog)
  * [TODO](#todo)
  * [Disclaimer](#disclaimer)

## Why use Wireguard-Ligase?

Main reasons behind using this script:

* Bash is installed and configured as a default shell on nearly any Linux distro.
* As of July 9th 2019 this script supports Arch, CentOS, Debian, Fedora, Manjaro, and Ubuntu
  * The script will check if it's running on a supported OS, then check if WireGuard is already installed.
  * If WireGuard is not installed, the script will offer to install it.
  * For Arch, Debian, Fedora, Manjaro, and Ubuntu - the script will choose an appropriate way to save netfilter rules in order for those to persist after reboot. See [Changelog](#changelog) for details.
* You are in control
  * Any system changes are shown to the user and user confirmation is required BEFORE any changes are made.
* Easily readable
  * It's easy to understand the inner workings of the script because bash is a very popular language.
  * Annotations will be added to obscure parts of the script to clarify the logic behind certain operations.

## Using the script

Using the script is fairly simple. There are a number of requirements for running the script and making everything work.

### Pre-requisites

1.  A server with a public IP address, either physical or cloud (DigitalOcean, LightSail, Vultr, etc), running Linux.

2.  The script will check if WireGuard is installed on Arch, CentOS, Debian, Fedora, Manjaro and Ubuntu. On other distributions, WireGuard must be installed BEFORE running the script. See [Installing Wireguard](#installing-wireguard) for more info.

3.  A user with sudo privileges for making any system changes.

### Usage

As mentioned before, using the script is very simple.
```bash
$ git clone https://github.com/SirToffski/WireGuard-Ligase.git # (1) Clone the repository.
$ cd WireGuard-Ligase/ # (2) cd into the repository.
$ sudo bash configure-wireguard.sh # (3) Run the script.
```

The script will guide you through the rest of the process.

## Installing Wireguard

The script has an ability to detect and install Wireguard of the following distributions:
* Arch Linux
* CentOS
* Debian
* Fedora
* Manjaro
* Ubuntu
* FreeBSD


For installation on other Operating Systems, please check the [**Wiki**](https://github.com/SirToffski/WireGuard-Ligase/wiki/Getting-Started).

## Changelog

See [**changelog**](/Changelog.md).

## TODO

  - [x] Add an option for a quick hands-free server deployment / host configuration. Especially useful for people who dont want to answer a lot of questions and are not too fussy about naming clients, etc.
  - [x] Customize server config depending on the distro in use (Arch vs Ubuntu, vs RHEL, etc)
  - [ ] Add option to configure other UNIX OS types (such as OpenBSD) - work in progress. 
    * FreeBSD support has been implemented.

## Disclaimer

The plan is to keep this updated and add features until the project reaches its logical conclusion.

Everyone is welcome to fork / contribute / use any or all parts of the project.

---
<sub><sub><sub><sub>WireGuard is a registered trademark of Jason A. Donenfeld.</sub></sub></sub></sub>



Changelog.md
## Changelog

* April 24th, 2019
  * Script to deploy the server would exit with an error if $pwd/keys/ and $pwd/client_confings/ did not exist during key creation.Logic to pre-check if the directory exists and to create one if needed was added. Script is now fully functional.
* April 29th, 2019
  * Similar pre-check as the above was added to the scipt to configure clients only.
  * All bash was updated with a new shebang.
  * New image generated with Carbon-cli instead of a screenshot.
  * Text colours and styling were added to improve readability.
* May 6th, 2019
  * Bugfixes:
    * Client config portion of the script had a syntax error reading `AllowedIPd` instead of `AllowedIPs`. This has been corrected.
    * Option to enable IP forwarding would change the value to `net.ipv4.ip_forward=1`, but would not uncomment it. This was fixed.
  * New features:
    * An option to install iptables-persistent and to enable systemctl service was added.
    * An option to enable WireGuard tunnel interface and to enable the interface on boot was added.
* May 20th, 2019
  * New Features:
    * Option for a quick-setup was added!
* July 9th, 2019
  * The script now checks if running as root.
  * New features:
    * Public IP of the machine on which the script is runnign is now automatically fetched via AWS `curl https://checkip.amazonaws.com`
    * Script now also checks if it's running on a supported OS and whether WireGuard is installed. If WireGuard is not installed, the script will offer to install it.
    * The script now customizes firewall rules based on the disstribution. On CentOS `firewall-cmd` will ne used. On Arch, Debian, Fedora, Manjaro and Ubuntu netfilter `iptables` are used. Furthermore, for the latter distributions, the script chooses an appropriate way to save netfilter rules. For example, on Debian and Ubuntu, `iptables-persistent` will be installed. On Fedora, netfilter rules are saved with `/sbin/service iptables save`. Finally, on Arch and Manjaro, systemd `iptables.service` is used with configuration saved to `/etc/iptables/iptables.rules`.
* July 22nd, 2019
  * Tons of minor bug fixes and some re-writes to save config logic
  * The script has been officially tested on CentOS 7 running in EC2 for the first time. Tons of minor adjustments had to be made.
    * CentOS 7 in EC2 does not have `firewalld` pre-installed and instead uses `iptables`. The iptables script has been written to take this into account. If OS type is CentOS and `firewalld` is installed, then `firewall-cmd` commands are used. If `firewalld` is not installed, `iptables` will be used by default. Also, the user will be offered to have `iptables-service` installed and enabled via `systemd` to have the rules persist after reboot.
* Oct 10th, 2019
  * It's been a while since the changelog has been updated; yet the work hasn't stopped.
  * Most of the commits were to improve user experience.
    * The look of the script has changed somewhat. The terminal screen is now cleared when needed and important parts summarized - easier to see than to explain.. but it's a lot better now.
    * Finally a good solution has been found to check for the directory where the script is running. See commit [7d2d61c](https://github.com/SirToffski/WireGuard-Ligase/commit/7d2d61c61949089a6b4aa363e422a5d53ac0423f). Find worked well on a brand new OS installation. However on a system with plenty of files and directories - it was a mess.
    * For the time being, work will continue concentrating on improvements to user interface, experience, and overall stability. New features will be implemented at a slower pace until the author is satisfied with UX.
* Jan 8th, 2020
  * Added initial support for FreeBSD.
* Jan 9th, 2020
  * BSD firewall ruless did not enable gateway mode in `/etc/rc.conf` and TCP segmentaion offloading in `/etc/sysctl.conf` - hence those settings were not preserved with server reboot. Both were fixed.
  * Using normal server setup mode now exports most of the variables to `shared_vars.sh`. This will prevent having to ask repetitive questions when setting up firewall rules.
* Jan 12th, 2020
  * FreeBSD support has been added to quick-setup mode.
  * Various syntax improvements in the shell code.

configure-wireguard.sh
#!/usr/bin/env bash

#   __          _______   _      _
#   \ \        / / ____| | |    (_)
#    \ \  /\  / / |  __  | |     _  __ _  __ _ ___  ___
#     \ \/  \/ /| | |_ | | |    | |/ _` |/ _` / __|/ _ \
#      \  /\  / | |__| | | |____| | (_| | (_| \__ \  __/
#       \/  \/   \_____| |______|_|\__, |\__,_|___/\___|
#                                   __/ |
#                                  |___/
# This is a simple bash script to aid in configuring WireGuard tunnels and clients.
# Q: Hasn't this been done bedofe?
# A: Probably.
# Q: Why another WG configuration script?
# A: Why not?
# Q: ....Ligase???
# A: In biochemistry, a ligase is an enzyme that can catalyze the joining of two large molecules by forming a new chemical bond -
# - https://en.wikipedia.org/wiki/Ligase (May 19th, 2019)

if [ "$EUID" -ne 0 ]; then
  printf %s\\n "Please run the script as root."
  exit 1
fi

my_wgl_folder="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"

source "$my_wgl_folder"/doc/functions.sh
# Setting the colours function
colours

logo=$(cat "$my_wgl_folder"/doc/ascii-logo)
printf '\e[2J\e[H'
printf %b\\n "${BP}$logo${Off}"

printf %b\\n "${BW}Welcome to WG Ligase${Off}.

The script will guide you through the installaton process, allowing to
choose a starting point. The idea is for this script to be equally
suitable for new deployments, as well as for configuring a live
deployment.
"

printf %b\\n "
Let's begin. Please select from one of the following options:
-----------------------------------

${BW}1 = Normal Setup:${Off} I would like to configure a new server and clients from scratch.

-----------------------------------

${BW}2 = Quick Setup:${Off} You will only be asked to specify public server IP.${Off}

-----------------------------------

${BW}3 = Clients only:${Off} I just need to generate some client configs and add those to an existing server.${Off}

-----------------------------------

${BW}4 = Firewall:${Off} I just need commands to configure IPTABLEs/firewalld.

----------------------------------"

read -r -p "Option #: " scope_of_script

case "$scope_of_script" in
"1")
  sudo bash "$my_wgl_folder"/Scripts/deploy_new_server.sh
  ;;
"2")
  sudo bash "$my_wgl_folder"/Scripts/quick_setup.sh
  ;;
"3")
  sudo bash "$my_wgl_folder"/Scripts/client_config.sh
  ;;
"4")
  sudo bash "$my_wgl_folder"/Scripts/setup_iptables.sh
  ;;
*)
  printf %b\\n "${BR}Sorry, wrong choise. Rerun the script and try again${Off}"
  ;;
esac


Scripts/deploy_new_server.sh
#!/usr/bin/env bash

if [ "$EUID" -ne 0 ]; then
  printf %b\\n "Please run the script as root."
  exit 1
fi

printf %s\\n "+--------------------------------------------+"
# Default working directory of the script.
## Requirements: Cloning the entire repository.
my_wgl_folder="$(cd "$(dirname "${BASH_SOURCE[0]}")" && cd .. >/dev/null 2>&1 && pwd)"
# A simple check if the entire repo was cloned.
## If not, working directory is a directory of the currently running script.
check_for_full_clone="$my_wgl_folder/configure-wireguard.sh"
if [ ! -f "$check_for_full_clone" ]; then
  my_wgl_folder="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
else
  source "$my_wgl_folder"/doc/functions.sh
  # Setting the colours function
  colours
fi

# Create a file with shared variables between the scripts
printf %b\\n "#!/usr/bin/env bash" >"$my_wgl_folder"/shared_vars.sh

# Determine the public IP of the host.
check_pub_ip=$(curl -s https://checkip.amazonaws.com)

######################## Pre-checks ##############################
# Check if a directory /keys/ exists, if not, it will be made
check_for_keys_directory=$("$my_wgl_folder"/keys)
if [ ! -d "$check_for_keys_directory" ]; then
  mkdir -p "$my_wgl_folder"/keys
fi

# Check if a directory /client_configs/ exists, if not, it will be made
check_for_clients_directory=$("$my_wgl_folder"/client_configs)

if [ ! -d "$check_for_clients_directory" ]; then
  mkdir -p "$my_wgl_folder"/client_configs
fi
##################### Pre-checks finished #########################

printf '\e[2J\e[H'

printf %b\\n "This script will take you through the steps needed to deploy a new server
and configure some clients."

if [ -f "$check_for_full_clone" ]; then
  printf %b\\n "\n First, let's check if wireguard is installed..."

  ############## Determine OS Type ##############
  # see /doc/functions.sh for more info
  ###############################################
  determine_os
  ############### FINISHED CHECKING OS AND OFFER TO INSTALL WIREGUARD ###############
fi

# Private address could be any address within RFC 1918,
# usually the first useable address in a /24 range.
# This however is completely up to you.
printf %b\\n "\n ${BW}Step 1)${Off} ${IW}Please specify the private address of the WireGuard server.${Off}"
read -r -p "Address: " server_private_range

printf '\e[2J\e[H'

# This would be a UDP port the WireGuard server would listen on.
printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
+--------------------------------------------+
\n${BW}Step 2)${Off} ${IW}Please specify listen port of the server.${Off}\n"
read -r -p "Listen port: " server_listen_port

printf '\e[2J\e[H'

# Public IP address of the server hosting the WireGuard server
printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
${BW}Server listen port = ${BR}$server_listen_port${Off}
+--------------------------------------------+
\n${BW}Step 3)${Off} ${IW}The public IP address of this machine is $check_pub_ip. 
Is this the address you would like to use? ${Off}
\n${BW}1 = yes, 2 = no${Off}"
read -r -p "Choice: " public_address

printf %s\\n "+--------------------------------------------+"

if [ "$public_address" = 1 ]; then
  server_public_address="$check_pub_ip"
elif [ "$public_address" = 2 ]; then
  printf %b\\n "\n${IW}Please specify the public address of the server.${Off}"
  read -r -p "Public IP: " server_public_address
  printf %s\\n "+--------------------------------------------+"
fi

printf '\e[2J\e[H'

# Internet facing iface of the server hosting the WireGuard server
printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
${BW}Server listen port = ${BR}$server_listen_port${Off}
${BW}Server public address = ${BR}$server_public_address${Off}
+--------------------------------------------+"
printf %b\\n "\n${BW}Step 4)${IW} Please also provide the internet facing interface of the server. 
${BW}Example: ${BR}eth0${Off}"
if [ "$distro" != "freebsd" ]; then
  printf %b\\n "\n Available interfaces are:
+--------------------+
$(ip -br a | awk '{print $1}')
+--------------------+"
else
  printf %b\\n "\n Available interfaces are:
+--------------------+
$(ifconfig -l)
+--------------------+"
fi

read -r -p "Interface: " local_interface

printf '\e[2J\e[H'

printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
${BW}Server listen port = ${BR}$server_listen_port${Off}
${BW}Server public address = ${BR}$server_public_address${Off}
${BW}WAN interface = ${BR}$local_interface${Off}
+--------------------------------------------+ \n"

read -n 1 -s -r -p "
Review the above. 
Press any key to continue 
Press r/R to try again
Press e/E to exit
" your_choice

case "$your_choice" in
[Rr]*)
  sudo bash "$my_wgl_folder"/Scripts/deploy_new_server.sh
  ;;
[Ee]*)
  exit
  ;;
*)
  printf '\e[2J\e[H'
  ;;
esac

printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
${BW}Server listen port = ${BR}$server_listen_port${Off}
${BW}Server public address = ${BR}$server_public_address${Off}
${BW}WAN interface = ${BR}$local_interface${Off}
+--------------------------------------------+"

# This would be the private and public keys of the server.
# If you are using this script, chances are those have not yet been generated yet.

printf %b\\n "\n${IW}Do you need to generate server keys?${Off} 
(If you have not yet configured the server, the probably yes).

${BW}1 = yes, 2 = no${Off}\n"

read -r -p "Choice: " generate_server_key
printf %s\\n "+--------------------------------------------+"

if [ "$generate_server_key" = 1 ]; then
  wg genkey | tee "$my_wgl_folder"/keys/ServerPrivatekey | wg pubkey >"$my_wgl_folder"/keys/ServerPublickey
  chmod 600 "$my_wgl_folder"/keys/ServerPrivatekey && chmod 600 "$my_wgl_folder"/keys/ServerPublickey

# The else statement assumes the user already has server keys,
# hence the option to generate them was not chosen.
# For the script to generate a server config, the user is asked
# to provide public/private key pair for the server.

else
  printf %b\\n "\n${IW}Specify server private key.${Off}\n"
  read -r -p "Server private key: " server_private_key
  printf %b\\n "$server_private_key" >"$my_wgl_folder"/keys/ServerPrivatekey
  printf %s\\n "+--------------------------------------------+"
  printf %b\\n "\n${IW}Specify server public key.${Off}\n"
  read -r -p "Server public key: " server_public_key
  printf %b\\n "$server_public_key" >"$my_wgl_folder"/keys/ServerPublickey
  chmod 600 "$my_wgl_folder"/keys/ServerPrivatekey && chmod 600 "$my_wgl_folder"/keys/ServerPrivatekey
  printf %s\\n "+--------------------------------------------+"

fi

sever_private_key_output=$(cat "$my_wgl_folder"/keys/ServerPrivatekey)
sever_public_key_output=$(cat "$my_wgl_folder"/keys/ServerPublickey)

printf %b\\n "\n${IW}Specify wireguard server interface name 
(will be the same as config name, without .conf)${Off}\n"

read -r -p "WireGuard Interface: " wg_serv_iface

printf '\e[2J\e[H'

printf %b\\n "\n+--------------------------------------------+
${BW}Server private address = ${BR}$server_private_range${Off}
${BW}Server listen port = ${BR}$server_listen_port${Off}
${BW}Server public address = ${BR}$server_public_address${Off}
${BW}WAN interface = ${BR}$local_interface${Off}
${BW}WireGuard interface = ${BR}$wg_serv_iface${Off}
+--------------------------------------------+\n"

{
  printf %b\\n "server_private_range=$server_private_range"
  printf %b\\n "server_listen_port=$server_listen_port"
  printf %b\\n "server_public_address=$server_public_address"
  printf %b\\n "local_interface=$local_interface"
  printf %b\\n "wg_serv_iface=$wg_serv_iface"
} >>"$my_wgl_folder"/shared_vars.sh

printf %b\\n "\n Generating server config file...."

sleep 2

if [ "$distro" != "freebsd" ]; then

  new_server_config=$(printf %b\\n "
[Interface]
Address = $server_private_range/32
SaveConfig = true
PostUp = iptables -A FORWARD -i $wg_serv_iface -j ACCEPT; iptables -t nat -A POSTROUTING -o $local_interface -j MASQUERADE; ip6tables -A FORWARD -i $wg_serv_iface -j ACCEPT; ip6tables -t nat -A POSTROUTING -o $local_interface -j MASQUERADE
PostDown = iptables -D FORWARD -i $wg_serv_iface -j ACCEPT; iptables -t nat -D POSTROUTING -o $local_interface -j MASQUERADE; ip6tables -D FORWARD -i $wg_serv_iface -j ACCEPT; ip6tables -t nat -D POSTROUTING -o $local_interface -j MASQUERADE
ListenPort = $server_listen_port
PrivateKey = $sever_private_key_output
  ")

else
  # We wont use iptables in FreeBSD, everythig will be handled by IPFW.
  new_server_config=$(printf %b\\n "
[Interface]
Address = $server_private_range/32
SaveConfig = true
ListenPort = $server_listen_port
PrivateKey = $sever_private_key_output
  ")

fi

printf %b\\n "$new_server_config" >"$my_wgl_folder"/"$wg_serv_iface".conf
chmod 600 "$my_wgl_folder"/"$wg_serv_iface".conf

printf %b\\n "Server config has been written to a file $my_wgl_folder/$wg_serv_iface.conf"
printf %s\\n "+--------------------------------------------+"

sleep 2

printf %b\\n "\n ${IW}Save config to /etc/wireguard/?${Off}\n
NOTE: ${UW}Choosing to save the config under the same file-name as
an existing config will ${BR}overrite it.${Off}\n
This script will check if a config file with the same name already
exists. It will back the existing config up before overriting it.
+--------------------------------------------+\n
Save config: ${BW}1 = yes, 2 = no${Off}\n"

check_for_existing_config="/etc/wireguard/$wg_serv_iface.conf"

read -r -p "Choice: " save_server_config

# The if statement checks whether a config with the same filename already exists.
# If it does, the falue will always be less than zero, hence it needs to be backed up.
if [ "$save_server_config" = 1 ] && [ -f "$check_for_existing_config" ]; then
  printf %b\\n "
  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    Found existing config file with the same name. 
    Backing up to /etc/wireguard/$wg_serv_iface.conf.bak
  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
  sleep 2
  mv /etc/wireguard/"$wg_serv_iface".conf /etc/wireguard/"$wg_serv_iface".conf.bak
  sleep 1
  printf %b\\n "$new_server_config" >/etc/wireguard/"$wg_serv_iface".conf
  printf '\e[2J\e[H'
  printf %b\\n "\nCongrats! Server config is ready and saved to \n/etc/wireguard/$wg_serv_iface.conf... The config is shown below."
elif [ "$save_server_config" = 1 ] && [ ! -f "$check_for_existing_config" ]; then
  # Make /etc/wireguard if it does not exist yet
  # Example is FreeBSD - /etc/wireguard is not automatically created
  # after installing WG.
  mkdir -p /etc/wireguard
  printf %b\\n "$new_server_config" >/etc/wireguard/"$wg_serv_iface".conf
  printf '\e[2J\e[H'
  printf %b\\n "\nCongrats! Server config is ready and saved to \n/etc/wireguard/$wg_serv_iface.conf... The config is shown below."
elif [ "$save_server_config" = 2 ]; then
  printf '\e[2J\e[H'
  printf %b\\n "\nUnderstood! Server config copy \nis located in $my_wgl_folder/$wg_serv_iface.conf.\nThe config is shown below."
fi

printf %b\\n "\n\n${IY}$new_server_config${Off}\n" | sed -E 's/PrivateKey = .*/PrivateKey = Hidden/g'

printf %s\\n "+--------------------------------------------+"

printf %b\\n "\n${IW}Configure clients?${Off}
${BW}1=yes, 2=no${Off}"

read -r -p "Choice: " client_config_answer
printf %s\\n "+--------------------------------------------+"

if [ "$client_config_answer" = 1 ]; then
  printf '\e[2J\e[H'
  printf %b\\n "\n${IW}How many clients would you like to configure?${Off}\n"
  read -r -p "Number of clients: " number_of_clients

  printf %s\\n "+--------------------------------------------+"

  printf %b\\n "\n${IW}Specify the DNS server your clients will use.${Off}\n"
  # This would usually be a public DNS server, for example 1.1.1.1,
  # 8.8.8.8, etc.
  read -r -p "DNS server: " client_dns
  printf '\e[2J\e[H'
  printf %b\\n "\nNext steps will ask to provide \nprivate address and a name for each client, one at a time.\n"
  printf %s\\n "+--------------------------------------------+"

  # Private address would be within the RFC 1918 range of the server.
  # For example if the server IP is 10.10.10.1/24, the first client
  # would usually have an IP of 10.10.10.2; though this can be any
  # address as long as it's within the range specified for the server.
  while read -r i; do
    printf %b\\n "\n${IW}Private address of client # $i (do NOT include /32):${Off}\n"
    read -r -p "Client $i IP: " client_private_address_["$i"]
    # Client name can be anything, mainly to easily identify the device
    # to be used. Some exampmles are:
    # Tom_iPhone
    # Wendy_laptop

    printf %s\\n "+--------------------------------------------+"

    printf %b\\n "\n${IW}Provide the name of the client # $i ${Off}\n"
    read -r -p "Client $i name: " client_name_["$i"]

    printf %s\\n "+--------------------------------------------+"

    wg genkey | tee "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey | wg pubkey >"$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey

    chmod 600 "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey
    chmod 600 "$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey

    client_private_key_["$i"]="$(cat "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey)"
    client_public_key_["$i"]="$(cat "$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey)"

    printf %b\\n "\n[Interface]
Address = ${client_private_address_["$i"]}
PrivateKey = ${client_private_key_["$i"]}
DNS = $client_dns\n
[Peer]
PublicKey = $sever_public_key_output
Endpoint = $server_public_address:$server_listen_port
AllowedIPs = 0.0.0.0/0
PersistentKeepalive = 21" >"$my_wgl_folder"/client_configs/"${client_name_["$i"]}".conf
    printf '\e[2J\e[H'
  done < <(seq 1 "$number_of_clients")
  printf %b\\n "\nAwesome!\nClient config files were saved to ${IW}$my_wgl_folder/client_configs/${Off}"
else
  printf %s\\n "+--------------------------------------------+"
  printf %b\\n "${IW}Before ending this script,\nwould you like to setup firewall rules for the new server? (recommended)${Off}\n
  ${BW}1 = yes, 2 = no${Off}\n"
  read -r -p "Choice: " iptables_setup
  if [ "$iptables_setup" = 1 ]; then
    sudo bash "$my_wgl_folder"/Scripts/setup_iptables.sh
  else
    printf %b\\n "Sounds good. Ending the scritp..."
    exit
  fi
fi
printf %b\\n "\n${IW}If you've got qrencode installed, the script can generate QR codes for
the client configs.\n\n Would you like to have QR codes generated?
\n1= yes, 2 = no${Off}"

read -r -p "Choice: " generate_qr_code

if [ "$generate_qr_code" = 1 ]; then
  while read -r q; do
    printf %b\\n "${BR}${client_name_[$q]}${Off}\n"
    qrencode -t ansiutf8 <"$my_wgl_folder"/client_configs/"${client_name_["$q"]}".conf
    printf %s\\n "+--------------------------------------------+"
  done < <(seq 1 "$number_of_clients")
elif [ "$generate_qr_code" = 2 ]; then
  printf %b\\n "\nAlright.. Moving on!\n+--------------------------------------------+"
else
  printf %b\\n "Sorry, wrong choice! Moving on with the script."
fi

printf %b\\n "\n${IW}Would you like to add client info to the server config now?${Off}
\n${BW}1 = yes, 2 = no${Off}"
read -r -p "Choice: " configure_server_with_clients

# If you chose to add client info to the server config AND to save the server config
# to /etc/wireguard/, then the script will add the clients to that config
if [ "$configure_server_with_clients" = 1 ]; then
  while read -r a; do
    printf %b\\n "\n[Peer]
PublicKey = ${client_public_key_["$a"]}
AllowedIPs = ${client_private_address_["$a"]}/32\n" >>/etc/wireguard/"$wg_serv_iface".conf
  done < <(seq 1 "$number_of_clients")
elif [ "$configure_server_with_clients" = 2 ]; then
  printf %b\\n "\nAlright, you may add the following to a server config file to setup clients.
\n-----------------\n"
  while read -r d; do
    printf %b\\n "\n${IY}[Peer]
PublicKey = ${client_public_key_["$d"]}
AllowedIPs = ${client_private_address_["$d"]}/32${Off}\n"
  done < <(seq 1 "$number_of_clients")
fi

printf %b\\n "-----------------"

# This assumes the WireGuard is already installed on the server.
# The script checks is there is config in /etc/wireguard/, if there is one,
# the value of the grep will be greater than or equal to 1, means it can be used
# to bring up the WireGuard tunnel interface.
printf %b\\n "${IW}Almost done!
Would you like to bring WireGuard interface up and to enable the service on boot?${Off}
\n${BW}1 = yes, 2 = no${Off}\n"

read -r -p "Choice: " enable_on_boot
printf '\e[2J\e[H'
if [ "$enable_on_boot" = 1 ]; then
  # If current OS is FreeBSD - we wont use systemd as we would've for supported linux distros.
  freebsd_os=$(uname -a | awk '{print $1}' | grep -i -c FreeBSD)
  if [ "$freebsd_os" -gt 0 ]; then
    printf %b\\n "\n${IY}chown -v root:root /etc/wireguard/$wg_serv_iface.conf
chmod -v 600 /etc/wireguard/$wg_serv_iface.conf
sysrc wireguard_enable=\"YES\"
sysrc wireguard_interfaces=\"$wg_serv_iface\"
service wireguard start${Off}\n"

    read -n 1 -s -r -p "
  Review the above. 
  Press any key to continue 
  Press r/R to restart the script
  Press e/E to exit
  " your_choice

    case "$your_choice" in
    [Rr]*)
      sudo bash "$my_wgl_folder"/Scripts/deploy_new_server.sh
      ;;
    [Ee]*)
      exit
      ;;
    *)
      chown -v root:root /etc/wireguard/"$wg_serv_iface".conf
      chmod -v 600 /etc/wireguard/"$wg_serv_iface".conf
      sysrc wireguard_enable="YES"
      sysrc wireguard_interfaces="$wg_serv_iface"
      service wireguard start
      ;;
    esac

  else
    printf %b\\n "\n${IY}chown -v root:root /etc/wireguard/$wg_serv_iface.conf
  chmod -v 600 /etc/wireguard/$wg_serv_iface.conf
  wg-quick up $wg_serv_iface
  systemctl enable wg-quick@$wg_serv_iface.service${Off}\n"

    read -n 1 -s -r -p "
  Review the above. 
  Press any key to continue 
  Press r/R to restart the script
  Press e/E to exit
  " your_choice

    case "$your_choice" in
    [Rr]*)
      sudo bash "$my_wgl_folder"/Scripts/deploy_new_server.sh
      ;;
    [Ee]*)
      exit
      ;;
    *)
      chown -v root:root /etc/wireguard/"$wg_serv_iface".conf
      chmod -v 600 /etc/wireguard/"$wg_serv_iface".conf
      wg-quick up "$wg_serv_iface"
      systemctl enable wg-quick@"$wg_serv_iface".service
      ;;
    esac
  fi
elif [ "$enable_on_boot" = 2 ]; then
  printf %b\\n "\n${IW} To manually enable the service and bring tunnel interface up,
  the following commands can be used:${Off}"
  if [ "$freebsd_os" = 0 ]; then
    printf %b\\n "\n${IY}chown -v root:root /etc/wireguard/$wg_serv_iface.conf
chmod -v 600 /etc/wireguard/$wg_serv_iface.conf
wg-quick up $wg_serv_iface
systemctl enable wg-quick@$wg_serv_iface.service${Off}"
  else
    printf %b\\n "\n${IY}chown -v root:root /etc/wireguard/$wg_serv_iface.conf
chmod -v 600 /etc/wireguard/$wg_serv_iface.conf
sysrc wireguard_enable=\"YES\"
sysrc wireguard_interfaces=$wg_serv_iface
service wireguard start${Off}"
  fi
fi

printf %b\\n "\n${IW}Before ending this script, would you like to setup firewall rules for the new server? (recommended)${Off}
\n${BW}1 = yes, 2 = no${Off}\n"

read -r -p "Choice: " iptables_setup
if [ "$iptables_setup" = 1 ]; then
  sudo bash "$my_wgl_folder"/Scripts/setup_iptables.sh
else
  printf %b\\n "Sounds good. Ending the script..."
fi
exit


Scripts/client_config.sh
#!/usr/bin/env bash

if [ "$EUID" -ne 0 ]; then
  printf %b\\n "Please run the script as root."
  exit 1
fi

# A simple check if the entire repo was cloned.
## If not, working directory is a directory of the currently running script.
check_for_full_clone="$my_wgl_folder/configure-wireguard.sh"
if [ ! -f "$check_for_full_clone" ]; then
  my_wgl_folder="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
else
  source "$my_wgl_folder"/doc/functions.sh
  # Setting the colours function
  colours
fi
######################## Pre-checks ##############################
# Check if a directory /keys/ exists, if not, it will be made
check_for_keys_directory=$("$my_wgl_folder"/keys)
if [ ! -d "$check_for_keys_directory" ]; then
  mkdir -p "$my_wgl_folder"/keys
fi

# Check if a directory /client_configs/ exists, if not, it will be made
check_for_clients_directory=$("$my_wgl_folder"/client_configs)

if [ ! -d "$check_for_clients_directory" ]; then
  mkdir -p "$my_wgl_folder"/client_configs
fi
##################### Pre-checks finished #########################

printf %b\\n "This script will help in easily generating client config for WireGuard."

printf %b\\n "\nWhat is the server's public key?\n"

read -r sever_public_key_output

printf %b\\n "\nWhat is the server listen port and public IP address?
\nExample: If public IP is 1.1.1.1 and listet port is 9201 -- type 1.1.1.1:9201"

read -r server_details

printf %b\\n "\nHow many clients would you like to configure?\n"
read -r number_of_clients
printf %b\\n "\nSpecify the DNS server your clients will use.\n"
read -r client_dns
printf %b\\n "\nNext steps will ask to provide private address and a name for each client, one at a time.\n"
while read -r i; do
  printf %b\\n "\nPrivate address of a client (do NOT include /32):\n"
  read -r client_private_address_["$i"]
  printf %b\\n "\nProvide the name of the client\n"

  read -r client_name_["$i"]

  wg genkey | tee "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey | wg pubkey >"$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey

  chmod 600 "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey
  chmod 600 "$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey

  client_private_key_["$i"]="$(cat "$my_wgl_folder"/keys/"${client_name_["$i"]}"Privatekey)"
  client_public_key_["$i"]="$(cat "$my_wgl_folder"/keys/"${client_name_["$i"]}"Publickey)"

  printf %b\\n "\n[Interface]
Address = ${client_private_address_["$i"]}
PrivateKey = ${client_private_key_["$i"]}
DNS = $client_dns
\n[Peer]
PublicKey = $sever_public_key_output
Endpoint = $server_details
AllowedIPs = 0.0.0.0/0
PersistentKeepalive = 21" >"$my_wgl_folder"/client_configs/"${client_name_["$i"]}".conf
done < <(seq 1 "$number_of_clients")

printf %b\\n "\nAwesome!\nClient config files were saved to $my_wgl_folder/client_configs/"

printf %b\\n "\nWould you link to add client info to the server config now?\n
1 = yes, 2 = no"
read -r configure_server_with_clients

if [ "$configure_server_with_clients" = 1 ]; then
  printf %b\\n "\nPlease specify the file name of the server config located in /etc/wireguard/ WITHOUT the  '.conf'.

  EXAMPLE:

  If your server config file is 'wg0.conf', type 'wg0'\n
  
  NOTE: You need to disable server WireGuard interface before changing the config."

  read -r server_file_for_clients
  while read -r c; do
    printf %b\\n "
[Peer]
PublicKey = ${client_public_key_["$c"]}
AllowedIPs = ${client_private_address_["$c"]}/32
  " | tee -a /etc/wireguard/"$server_file_for_clients".conf >/dev/null
  done < <(seq 1 "$number_of_clients")
else
  printf %b\\n "
  Alright, you may add the following to a server config file to setup clients.

  -----------------
  "
  while read -r d; do
    printf %b\\n "
[Peer]
PublicKey = ${client_public_key_["$d"]}
AllowedIPs = ${client_private_address_["$d"]}/32
"
  done < <(seq 1 "$number_of_clients")
  printf %b\\n "-----------------"
fi
